juju/juju
# Use a key-based mutex lock to download resources/agent bins once.

- Adds github.com/im7mortal/kmutex dep
- Adds global kmutex lock for github.com/juju/juju/resource/repositories.GetResource
- Changes agent from x/sync/singleflight to kmutex 

## QA steps

- bootstrap lxd with trace logging
- `juju bootstrap localhost`
- `juju model-config -m controller logging-config="<root>=TRACE"`
- `juju deploy test-resource-pull --channel=beta -n 10`
- wait for units to be ready
- `juju run-action test-resource-pull/0 test-resource-pull/1 test-resource-pull/2 test-resource-pull/3 test-resource-pull/4 test-resource-pull/5 test-resource-pull/6 test-resource-pull/7 test-resource-pull/8 test-resource-pull/9 pull`
- `juju debug-log -m controller --replay | grep GetResource`
- should only be one GetResource line

## Documentation changes

N/A

## Bug reference

https://bugs.launchpad.net/juju/+bug/1905703
